File: backend/app/agent/workflow.py
"""把 Harness Evidence 投影为模型可见状态，并约束下一批 Tool 调用。"""

from __future__ import annotations

import json
from dataclasses import dataclass
from enum import StrEnum
from typing import Any

from agents import RunHooks
from agents.run_config import CallModelData, ModelInputData
from agents.tool_guardrails import (
    ToolGuardrailFunctionOutput,
    ToolInputGuardrailData,
    tool_input_guardrail,
)
from openai.types.responses import ResponseFunctionToolCall

from app.agent.evidence import RecommendationEvidence


class WorkflowStage(StrEnum):
    NEED_SUPPORT = "need_support"
    READY_TO_DRAFT = "ready_to_draft"


@dataclass(frozen=True)
class GateDecision:
    allowed: bool
    reason: str | None = None


@dataclass(frozen=True)
class ToolBatch:
    stage: WorkflowStage
    decisions: dict[str, GateDecision]


def stage_for(evidence: RecommendationEvidence) -> WorkflowStage:
    """Evidence 是唯一状态源；阶段不单独持久化，避免两份状态漂移。"""

    used = set(evidence.used_tools)
    support = set(evidence.required_support_tools)
    if not support.issubset(used):
        return WorkflowStage.NEED_SUPPORT
    if not set(evidence.required_knowledge_scopes).issubset(
        evidence.completed_knowledge_scopes
    ):
        return WorkflowStage.NEED_SUPPORT
    return WorkflowStage.READY_TO_DRAFT


def allowed_tools(evidence: RecommendationEvidence) -> tuple[str, ...]:
    used = set(evidence.used_tools)
    allowed: list[str] = []
    if evidence.used_tools.count("retrieve_knowledge") < 3:
        allowed.append("retrieve_knowledge")
    if (
        "get_marketing_strategy" in evidence.required_support_tools
        and "get_marketing_strategy" not in used
    ):
        allowed.append("get_marketing_strategy")
    return tuple(allowed)


def plan_tool_batch(
    evidence: RecommendationEvidence,
    calls: list[tuple[str, str]],
) -> ToolBatch:
    """在 Tool 执行前冻结整批决策，避免并发调用依赖完成先后顺序。"""

    stage = stage_for(evidence)
    allowed = set(allowed_tools(evidence))
    accepted: set[str] = set()
    decisions: dict[str, GateDecision] = {}
    for call_id, tool_name in calls:
        if tool_name not in allowed:
            decisions[call_id] = GateDecision(False, "tool_not_allowed_in_stage")
        elif tool_name in accepted:
            decisions[call_id] = GateDecision(False, "duplicate_tool_in_batch")
        else:
            decisions[call_id] = GateDecision(True)
            accepted.add(tool_name)
    return ToolBatch(stage=stage, decisions=decisions)


def render_agent_status(evidence: RecommendationEvidence) -> str:
    """只披露进度与下一步，不把 Harness-owned Evidence 交给模型。"""

    stage = stage_for(evidence)
    completed = ", ".join(dict.fromkeys(evidence.used_tools)) or "none"
    required_scopes = ", ".join(evidence.required_knowledge_scopes) or "none"
    completed_scopes = ", ".join(sorted(evidence.completed_knowledge_scopes)) or "none"
    next_steps = list(allowed_tools(evidence))
    if stage is WorkflowStage.READY_TO_DRAFT:
        next_steps.append("final_output")
    allowed = ", ".join(next_steps)
    version = len(evidence.used_tools) + len(evidence.blocked_attempts)
    return (
        f'<agent_status version="{version}">\n'
        f"stage: {stage.value}\n"
        f"completed_steps: {completed}\n"
        f"required_knowledge_scopes: {required_scopes}\n"
        f"completed_knowledge_scopes: {completed_scopes}\n"
        f"allowed_next: {allowed}\n"
        f"blocked_attempts: {len(evidence.blocked_attempts)}\n"
        "</agent_status>"
    )


def append_agent_status(data: CallModelData[Any]) -> ModelInputData:
    """每轮重新追加状态快照，不改写原始对话 Context。"""

    if data.context is None:
        return data.model_data
    status_item: Any = {
        "role": "developer",
        "content": [
            {
                "type": "input_text",
                "text": render_agent_status(data.context.evidence),
            }
        ],
    }
    return ModelInputData(
        input=[*data.model_data.input, status_item],
        instructions=data.model_data.instructions,
    )


class ChattyRunHooks(RunHooks[Any]):
    async def on_llm_end(self, context: Any, agent: Any, response: Any) -> None:
        del agent
        calls = [
            (item.call_id, item.name)
            for item in response.output
            if isinstance(item, ResponseFunctionToolCall)
        ]
        batch = plan_tool_batch(context.context.evidence, calls)
        context.context.batch_stage = batch.stage
        context.context.batch_decisions = batch.decisions


@tool_input_guardrail(name="chatty_stage_gate")
async def stage_guardrail(
    data: ToolInputGuardrailData,
) -> ToolGuardrailFunctionOutput:
    """执行 Hook 已冻结的批次决策，并把拒绝原因返回模型。"""

    tool_context = data.context
    run_context = tool_context.context
    decision = run_context.batch_decisions.get(tool_context.tool_call_id)
    if decision is not None and decision.allowed:
        stage = run_context.batch_stage or stage_for(run_context.evidence)
        return ToolGuardrailFunctionOutput.allow({"stage_snapshot": stage.value})

    reason = "unplanned_tool_call"
    if decision is not None and decision.reason is not None:
        reason = decision.reason
    stage = run_context.batch_stage or stage_for(run_context.evidence)
    run_context.evidence.blocked_attempts.append(
        f"{stage.value}:{tool_context.tool_name}:{reason}"
    )
    message = json.dumps(
        {
            "status": "blocked",
            "reason": reason,
            "required_next": list(allowed_tools(run_context.evidence)),
            "agent_status": render_agent_status(run_context.evidence),
        },
        ensure_ascii=False,
    )
    return ToolGuardrailFunctionOutput.reject_content(
        message=message,
        output_info={"call_id": tool_context.tool_call_id},
    )

File: server/src/agent/workflow.ts
/**
 * 把 Harness Evidence 投影为模型可见状态，并约束下一批 Tool 调用。
 *
 * 主 Agent 可能一次提出多个 Tool call。这个文件在 Tool 真正执行前，根据 Evidence 为整批
 * 调用做决定：哪些允许、哪些重复、哪些当前阶段不能调用。决定会返回 Model，让它下一轮
 * 可以自行纠正；Harness 不依靠 Model 对自身调用历史的描述。
 */

import type { RecommendationEvidence } from "./evidence.ts";

/** Agent Loop 只有两个阶段：补齐支撑材料，或者已经可以生成草稿。 */
export const WorkflowStage = {
  NEED_SUPPORT: "need_support",
  READY_TO_DRAFT: "ready_to_draft",
} as const;
export type WorkflowStage = (typeof WorkflowStage)[keyof typeof WorkflowStage];

/** 单个 Tool call 的门禁结果。 */
export type GateDecision = {
  allowed: boolean;
  reason: string | null;
};

/** 同一次 Model 响应中全部 Tool call 的冻结决策。 */
export type ToolBatch = {
  stage: WorkflowStage;
  decisions: Map<string, GateDecision>;
};

/** Evidence 是唯一状态源；阶段不单独持久化，避免两份状态漂移。 */
export function stageFor(evidence: RecommendationEvidence): WorkflowStage {
  // 转成 Set 后只关心“是否完成”，不关心同一 Tool 调用了几次。
  const used = new Set(evidence.used_tools);
  if (!evidence.required_support_tools.every((tool) => used.has(tool))) {
    return WorkflowStage.NEED_SUPPORT;
  }
  if (
    !evidence.required_knowledge_scopes.every((scope) =>
      evidence.completed_knowledge_scopes.has(scope),
    )
  ) {
    return WorkflowStage.NEED_SUPPORT;
  }
  return WorkflowStage.READY_TO_DRAFT;
}

/** 返回当前 Evidence 状态下 Model 下一步可以调用的 Tool 名称。 */
export function allowedTools(evidence: RecommendationEvidence): string[] {
  const used = new Set(evidence.used_tools);
  const allowed: string[] = [];

  // 知识检索允许改写 query 重试，但整个 Agent Loop 最多三次。
  const knowledgeCalls = evidence.used_tools.filter(
    (tool) => tool === "retrieve_knowledge",
  ).length;
  if (knowledgeCalls < 3) allowed.push("retrieve_knowledge");
  if (
    evidence.required_support_tools.includes("get_marketing_strategy") &&
    !used.has("get_marketing_strategy")
  ) {
    allowed.push("get_marketing_strategy");
  }
  return allowed;
}

/** 在 Tool 执行前冻结整批决策，避免并发调用依赖完成先后顺序。 */
export function planToolBatch(
  evidence: RecommendationEvidence,
  calls: readonly (readonly [string, string])[],
): ToolBatch {
  // 在任何 Tool 开始前读取一次 Evidence，整批调用都基于同一个快照做决定。
  const stage = stageFor(evidence);
  const allowed = new Set(allowedTools(evidence));
  const accepted = new Set<string>();
  const decisions = new Map<string, GateDecision>();

  for (const [callId, toolName] of calls) {
    if (!allowed.has(toolName)) {
      decisions.set(callId, { allowed: false, reason: "tool_not_allowed_in_stage" });
    } else if (accepted.has(toolName)) {
      // 同一批重复 Tool 看不到彼此结果，所以只允许第一个执行。
      decisions.set(callId, { allowed: false, reason: "duplicate_tool_in_batch" });
    } else {
      decisions.set(callId, { allowed: true, reason: null });
      accepted.add(toolName);
    }
  }
  return { stage, decisions };
}

/**
 * 只披露进度与下一步，不把 Harness-owned Evidence 交给模型。
 *
 * 返回普通字符串而不是对象，因为它会作为 developer message 加入 Model Context。
 * 商品 ID、画像详情等真实 Evidence 不放进去，避免 Model 反过来篡改 Harness 判断。
 */
export function renderAgentStatus(evidence: RecommendationEvidence): string {
  const stage = stageFor(evidence);
  const completed = [...new Set(evidence.used_tools)].join(", ") || "none";
  const requiredScopes = evidence.required_knowledge_scopes.join(", ") || "none";
  const completedScopes =
    [...evidence.completed_knowledge_scopes].sort().join(", ") || "none";
  const nextSteps = allowedTools(evidence);
  if (stage === WorkflowStage.READY_TO_DRAFT) nextSteps.push("final_output");
  // version 不是数据库版本，只是让 Model 看出状态是否比上一轮有进展。
  const version = evidence.used_tools.length + evidence.blocked_attempts.length;
  return [
    `<agent_status version="${version}">`,
    `stage: ${stage}`,
    `completed_steps: ${completed}`,
    `required_knowledge_scopes: ${requiredScopes}`,
    `completed_knowledge_scopes: ${completedScopes}`,
    `allowed_next: ${nextSteps.join(", ")}`,
    `blocked_attempts: ${evidence.blocked_attempts.length}`,
    "</agent_status>",
  ].join("\n");
}
